### PR TITLE
feat(auth/oauth2adapt): adds a new module to translate types

### PR DESCRIPTION
**Pull request by codyoss**
This package can be used by us to make sure we preserve expected strongly typed errors in chains. It can also be used by customers and other libs to easily convert between the two equivalent interfaces between the new and old world libraries.

### CODE REVIEW

**Review by noahdietz — APPROVED** [review]

File: auth/oauth2adapt/oauth2adapt.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+// into a [cloud.google.com/go/auth.TokenProvider].
+func TokenProviderFromTokenSource(ts oauth2.TokenSource) auth.TokenProvider {
+	return &tokenProviderAdapter{ts: ts}
+}
+
+type tokenProviderAdapter struct {
+	ts oauth2.TokenSource
+}
+
+// Token fulfills the [cloud.google.com/go/auth.TokenProvider] interface.
```
Comment (quartzmo) [inline auth/oauth2adapt/oauth2adapt.go:38]: Maybe add a doc line about how this calls `Token` on the underlying `oauth2.TokenSource` and wraps the result?

File: auth/oauth2adapt/oauth2adapt.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+// [golang.org/x/oauth2.TokenSource].
+func TokenSourceFromTokenProvider(tp auth.TokenProvider) oauth2.TokenSource {
+	return &tokenSourceAdapter{tp: tp}
+}
+
+type tokenSourceAdapter struct {
+	tp auth.TokenProvider
+}
+
+// Token fulfills the [golang.org/x/oauth2.TokenSource] interface.
```
Comment (quartzmo) [inline auth/oauth2adapt/oauth2adapt.go:65]: Maybe add a doc line about how this calls Token on the underlying auth.TokenProvider and wraps the result?

File: auth/oauth2adapt/oauth2adapt.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+type oauth2Error struct {
+	ErrorCode        string `json:"error"`
+	ErrorDescription string `json:"error_description"`
+	ErrorURI         string `json:"error_uri"`
+}
+
+// AddRetrieveErrorToAuthError returns the same error provided and adds a
+// [golang.org/x/oauth2.RetrieveError] to the error chain by setting the `Err` field on the
+// [cloud.google.com/go/auth.Error].
+func AddRetrieveErrorToAuthError(err *auth.Error) *auth.Error {
```
Comment (quartzmo) [inline auth/oauth2adapt/oauth2adapt.go:90]: This func seems to have a little bit of logic, is it unit tested?

File: auth/oauth2adapt/oauth2adapt.go (hunk 1 above)
Comment (codyoss) [reply auth/oauth2adapt/oauth2adapt.go:38]: done.

File: auth/oauth2adapt/oauth2adapt.go (hunk 2 above)
Comment (codyoss) [reply auth/oauth2adapt/oauth2adapt.go:65]: done.

File: auth/oauth2adapt/oauth2adapt.go (hunk 3 above)
Comment (codyoss) [reply auth/oauth2adapt/oauth2adapt.go:90]: yup!

**Pull request merged**